### PR TITLE
kubeadm: Upload configuration used at 'kubeadm init' time to ConfigMap for easier upgrades

### LINKED ISSUE #373: Upload configuration used at `kubeadm init` time to a ConfigMap
And backport that to v1.7.4 for easier upgradability.

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

Uploads config used to a ConfigMap so we can upgrade a cluster seamlessly without forcing the user to re-specify all options they specified the first time.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes: https://github.com/kubernetes/kubeadm/issues/373

**Special notes for your reviewer**:

This should be a cherrypick-candidate for easier upgrading to v1.8
cc @wojtek-t 

**Release note**:

```release-note
kubeadm: Upload configuration used at 'kubeadm init' time to ConfigMap for easier upgrades
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @timothysc

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go
Comment (pipejakob) [inline cmd/kubeadm/app/cmd/init.go]: I have a bit of a nitpick about the verb here. "Save" is a little vague to me, and immediately brings up questions like "what happens if the config already exists? Will it get clobbered? Will it just save a new copy? Is it expected that this configuration was already created, and I'm just saving an update to it?"

I like how our other APIs are extremely informative about their contracts, like `tokenphase.CreateBootstrapConfigMapIfNotExists()` or `tokenphase.UpdateOrCreateToken()`. Can the phase be named something like `configphase` and the method be `CreateOrUpdateConfiguration()`? That also gives us the flexibility of also having other complementary APIs in the same phase like `CreateConfigIfNotExists()` in the future if there's demand for other variants.

File: cmd/kubeadm/app/cmd/phases/saveconfig.go
Comment (pipejakob) [inline cmd/kubeadm/app/cmd/phases/saveconfig.go]: node-name? Copy-paste error?

File: cmd/kubeadm/app/cmd/phases/saveconfig.go
Comment (pipejakob) [inline cmd/kubeadm/app/cmd/phases/saveconfig.go]: I don't know if we have this pattern for the other commands, but if we're not expecting anything in `args`, I would error in the case anything is provided. Someone might run `kubeadm phase saveconfig my-file` thinking that they're overriding the filename to use, but not realize that they forgot `--config` and accidentally use the wrong file.

File: cmd/kubeadm/app/phases/saveconfig/saveconfig.go
Comment (pipejakob) [inline cmd/kubeadm/app/phases/saveconfig/saveconfig.go]: Are there tests we can write for this? In particular, if we want this to be idempotent like our other phases so that `kubeadm init` can be used for upgrading, we should assert that `AlreadyExist` errors are handled properly, while other errors are terminal.

File: cmd/kubeadm/app/cmd/phases/saveconfig.go
Comment (jbeda) [inline cmd/kubeadm/app/cmd/phases/saveconfig.go]: I'm not sure of the name of the phase here.  Save sounds like it is absolutely necessary whereas this is more of a helper type of thing.  Perhaps "upload-config"?

Also, change the description a little to something like "Upload the currently used configuration for kubeadm to a ConfigMap in the cluster for future use in reconfiguration and upgrades of the cluster."

**Review by jbeda — APPROVED** [review]

**Comment by jbeda** [comment]
LGTM modulo @pipejakob's comments.

/approve

File: cmd/kubeadm/app/cmd/phases/saveconfig.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/phases/saveconfig.go]: Good catch, yeah, copy-paste error, should be `<config-file>` indeed

File: cmd/kubeadm/app/cmd/phases/saveconfig.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/phases/saveconfig.go]: I can add a comment about that as follow up. My node bootstrap token PR has a func that can do that.

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: I changed this to `UploadConfiguration` sounds good to you?

**Comment by mattmoyer** [comment]
/lgtm

I had some questions about using an object with a stricter schema instead of a ConfigMap, but @timothysc said that was already discussed and punted until later.

File: cmd/kubeadm/app/cmd/init.go
Comment (pipejakob) [inline cmd/kubeadm/app/cmd/init.go]: To me, "upload" is just as vague as "save," since it isn't obvious what happens if the data already exists. We already have the pattern of explicitly saying "CreateOrUpdate" elsewhere in our phases, so I'm wondering why we don't want to follow that. If we're trying to have kubeadm taken seriously as a library / toolkit, I think having a consistent API experience is pretty important.

But, three different people had three different ideas for the naming in this PR :-). Since this is basically introducing a new kubeadm API, maybe we should cast the net a little wider and get a little bit of consensus on how to name it?

Or, if we already have precedent of "all of the phases are alpha anyway, so we can rename them whenever we want," maybe it's not so important to get it right the first time. I haven't been involved in the kubeadm adoption working group, so I'm not sure what the sentiment is around using these phases.

**Comment by pipejakob** [comment]
Same request for new unit tests for the new code. It will never be easier to add tests than it is right now.

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
> I had some questions about using an object with a stricter schema instead of a ConfigMap, but @timothysc said that was already discussed and punted until later.

As discussed on Slack; there won't be a mechanism in the API for validating that configuration you're uploading to the cluster is valid anytime soon.

> To me, "upload" is just as vague as "save," since it isn't obvious what happens if the data already exists. We already have the pattern of explicitly saying "CreateOrUpdate" elsewhere in our phases, so I'm wondering why we don't want to follow that. If we're trying to have kubeadm taken seriously as a library / toolkit, I think having a consistent API experience is pretty important.

I agree with you. This is still in the design phase and I love input here. Please create an issue @pipejakob 

> But, three different people had three different ideas for the naming in this PR :-). Since this is basically introducing a new kubeadm API, maybe we should cast the net a little wider and get a little bit of consensus on how to name it?
> Or, if we already have precedent of "all of the phases are alpha anyway, so we can rename them whenever we want," maybe it's not so important to get it right the first time. I haven't been involved in the kubeadm adoption working group, so I'm not sure what the sentiment is around using these phases.

Yes -- right. We're defining APIs here, but the hard thing is the timeline here and the fact that this PR is in a dependency chain for upgrades. Adding this functionality is not trivial in the chain of  ~10 PRs I have posted or am posting as we speak. Getting through the SQ is not trivial either for multiple PRs in a row.

So, my proposal is that you'd create an issue for your concerns. We will discuss things there and spawn a PR with a naming-only change as soon as we find a better name.

Until that, we're not blocking the upgrade chain, instead merging this now-ish lets us work in parallell.

> Same request for new unit tests for the new code. It will never be easier to add tests than it is right now.

Again -- totally agree. However, I'm gonna be away for 3.5-days-ish and school starts on Monday after that so I'd really want to get this merged now. I can promise to make an unit test for this before the next SIG meeting if you let me.

Does this at all sound reasonable? Sorry for the bad timing -- help here would be very appreciated.

**Review by pipejakob — APPROVED** [review]
/lgtm

**Pull request merged**
